spanner/value.go

```diff
@@ -2069,6 +2072,47 @@ func decodeValue(v *proto3.Value, t *sppb.Type, ptr interface{}, opts ...decodeO
 			return decodableType.decodeValueToCustomType(v, t, acode, atypeAnnotation, ptr)
 		}
 
+		// Check if the proto encoding is for an array of proto message or array of proto enum.
+		// The below scenario cannot be handled in a switch case statement because the pointer in this case
+		// would be an array of proto message or array of proto enum which are user defined types.
+		// Adding a generic switch case for these user defined types is not possible, hence we use proto encoding.
+		if code == sppb.TypeCode_ARRAY && (acode == sppb.TypeCode_PROTO ||
+			acode == sppb.TypeCode_ENUM || acode == sppb.TypeCode_INT64 ||
+			acode == sppb.TypeCode_BYTES) {
+			rv := reflect.ValueOf(ptr)
+			typ := rv.Type()
+			if typ.Kind() != reflect.Ptr {
+				return errNotAPointer(ptr)
+			}
+			if isNull {
+				break
+			}
+			etyp := typ.Elem().Elem()
+			switch acode {
+			case sppb.TypeCode_PROTO, sppb.TypeCode_BYTES:
```

Comment:
Please break it into smaller functions for better readability.  One for PROTO/BYTES and other for ENUM/INT64.